### PR TITLE
Add not connected equipments  

### PR DESCRIPTION
**Pull request by jamal-khey**

### CODE REVIEW

File: src/main/java/org/gridsuite/actions/server/ContingencyListService.java [hunk 1]
```diff
@@ -190,15 +192,22 @@ private List<ContingencyInfos> evaluateContingencyList(PersistentContingencyList
...
         notFoundElements.entrySet().stream()
                 .filter(stringSetEntry -> contingencies.stream().noneMatch(c -> c.getId().equals(stringSetEntry.getKey())))
-                .map(stringSetEntry -> new ContingencyInfos(stringSetEntry.getKey(), null, stringSetEntry.getValue()))
+                .map(stringSetEntry -> new ContingencyInfos(stringSetEntry.getKey(), null, stringSetEntry.getValue(), null))
                 .forEach(contingencyInfos::add);
 
         contingencies.stream()
-                .map(contingency -> new ContingencyInfos(contingency, notFoundElements.get(contingency.getId())))
-                .forEach(contingencyInfos::add);
+                .map(contingency -> {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/actions/server/ContingencyListService.java:201]: Why we do not handle disconnected elements like not found elements ?

File: src/test/java/org/gridsuite/actions/test/FiltersUtilsTest.java [hunk 2]
```diff
@@ -94,4 +94,18 @@ public void test2WTransfo() {
         assertFalse(FiltersUtils.transfoMatch(transfo2, List.of("FR")));
         assertFalse(FiltersUtils.transfoMatch(transfo2, List.of("ES", "ZA", "DE", "FR")));
     }
+
+    @Test
+    public void testDisconnectEquipment() {
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/actions/test/FiltersUtilsTest.java:99]: Remove this test
Add a test with `ContingencyInfos`

File: src/main/java/org/gridsuite/actions/server/dto/ContingencyInfos.java [hunk 3]
```diff
@@ -27,8 +27,9 @@ public class ContingencyInfos {
     private String id;
     private Contingency contingency;
     private Set<String> notFoundElements;
+    private Set<String> notConnectedElements;
 
-    public ContingencyInfos(Contingency contingency, Set<String> notFoundElements) {
+    public ContingencyInfos(Contingency contingency, Set<String> notFoundElements, Set<String> notConnectedElements) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/actions/server/dto/ContingencyInfos.java:32]: Oh not used !
Need a complete test !

**Review by SlimaneAmar — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/actions/server/dto/ContingencyInfos.java (hunk 3 above)
Comment (jamal-khey) [reply src/main/java/org/gridsuite/actions/server/dto/ContingencyInfos.java]: removed

File: src/main/java/org/gridsuite/actions/server/ContingencyListService.java [hunk 4]
```diff
@@ -189,20 +189,44 @@ private List<ContingencyInfos> evaluateContingencyList(PersistentContingencyList
...
         contingencies.stream()
-                .map(contingency -> new ContingencyInfos(contingency, notFoundElements.get(contingency.getId())))
+                .map(contingency -> createContingencyInfoWithDisconnects(contingency, network, notFoundElements.getOrDefault(contingency.getId(), null)))
                 .forEach(contingencyInfos::add);
 
         return contingencyInfos;
     }
 
+    private ContingencyInfos createContingencyInfoWithDisconnects(Contingency contingency, Network network, Set<String> notFoundElements) {
+        Set<String> disconnects = contingency.getElements().stream()
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/actions/server/ContingencyListService.java:207]: disconnectedElements

File: src/main/java/org/gridsuite/actions/server/ContingencyListService.java [hunk 5]
```diff
@@ -189,20 +189,44 @@ private List<ContingencyInfos> evaluateContingencyList(PersistentContingencyList
...
-        // we add all the contingencies that have only wrong ids
         notFoundElements.entrySet().stream()
                 .filter(stringSetEntry -> contingencies.stream().noneMatch(c -> c.getId().equals(stringSetEntry.getKey())))
-                .map(stringSetEntry -> new ContingencyInfos(stringSetEntry.getKey(), null, stringSetEntry.getValue()))
+                .map(stringSetEntry -> new ContingencyInfos(stringSetEntry.getKey(), null, stringSetEntry.getValue(), null))
                 .forEach(contingencyInfos::add);
 
         contingencies.stream()
-                .map(contingency -> new ContingencyInfos(contingency, notFoundElements.get(contingency.getId())))
+                .map(contingency -> createContingencyInfoWithDisconnects(contingency, network, notFoundElements.getOrDefault(contingency.getId(), null)))
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/actions/server/ContingencyListService.java:200]: Why getOrDefault ?

File: src/test/java/org/gridsuite/actions/server/ContingencyListControllerTest.java [hunk 6]
```diff
@@ -815,6 +822,32 @@ public void testExportContingenciesInfos() throws Exception {
...
+                        .contentType(APPLICATION_JSON))
+                .andExpect(status().isOk()).andReturn().getResponse().getContentAsString();
+
+        UUID contingencyListId = objectMapper.readValue(res, IdBasedContingencyList.class).getId();
+
+        mvc.perform(get("/" + VERSION + "/contingency-lists/contingency-infos/" + contingencyListId + "/export?networkUuid=" + NETWORK_UUID + "&variantId=" + VARIANT_ID_2)
+                        .contentType(APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andExpect(content().contentTypeCompatibleWith(APPLICATION_JSON))
+                .andExpect(content().json("[{\"id\":\"default\",\"contingency\":{\"id\":\"default\",\"elements\":[{\"id\":\"NHV1_NHV2_2\",\"type\":\"LINE\"},{\"id\":\"NHV1_NHV2_1\",\"type\":\"LINE\"}]},\"notFoundElements\":[\"TEST1\"],\"notConnectedElements\":[\"NHV1_NHV2_1\"]}]"))
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/actions/server/ContingencyListControllerTest.java:843]: Never use a raw json 
Use mapper instead

File: src/main/java/org/gridsuite/actions/server/ContingencyListService.java (hunk 4 above)
Comment (jamal-khey) [reply src/main/java/org/gridsuite/actions/server/ContingencyListService.java]: done

File: src/main/java/org/gridsuite/actions/server/ContingencyListService.java (hunk 5 above)
Comment (jamal-khey) [reply src/main/java/org/gridsuite/actions/server/ContingencyListService.java]: agree , get return null by default if not found

File: src/test/java/org/gridsuite/actions/server/ContingencyListControllerTest.java [hunk 7]
```diff
@@ -815,6 +824,34 @@ public void testExportContingenciesInfos() throws Exception {
                 .andExpect(status().isOk());
     }
 
+    @Test
+    public void testExportContingenciesInfosNotConnectedAndNotFound() throws Exception {
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/actions/server/ContingencyListControllerTest.java:828]: testExportContingenciesWithNotConnectedAndNotFoundElements

File: src/test/java/org/gridsuite/actions/server/ContingencyListControllerTest.java [hunk 8]
```diff
@@ -815,6 +824,34 @@ public void testExportContingenciesInfos() throws Exception {
...
+
+        Instant modificationDate = Instant.now();
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/actions/server/ContingencyListControllerTest.java:830]: Remove var ?

**Review by SlimaneAmar — APPROVED** [review]

File: src/main/java/org/gridsuite/actions/server/ContingencyListService.java [hunk 9]
```diff
@@ -189,20 +189,44 @@ private List<ContingencyInfos> evaluateContingencyList(PersistentContingencyList
...
+                .filter(contingencyElement -> {
+                    var connectable = network.getConnectable(contingencyElement.getId());
+                    return connectable != null && isDisconnected(connectable);
+                })
+                .map(ContingencyElement::getId)
+                .collect(Collectors.toSet());
+        return new ContingencyInfos(contingency.getId(), contingency, notFoundElements, disconnectedElements);
+    }
+
+    public static boolean isDisconnected(Connectable<?> connectable) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/actions/server/ContingencyListService.java:217]: private

**Review by SlimaneAmar — APPROVED** [review]

**Review by AAJELLAL — COMMENTED** [review]
Tests and code are ok except one typo to fix.

File: src/main/java/org/gridsuite/actions/server/ContingencyListService.java [hunk 10]
```diff
@@ -189,20 +189,43 @@ private List<ContingencyInfos> evaluateContingencyList(PersistentContingencyList
...
+                    return connectable != null && isDisconnected(connectable);
+                })
+                .map(ContingencyElement::getId)
+                .collect(Collectors.toSet());
+    }
+
+    private boolean isDisconnected(Connectable<?> connectable) {
+        List<? extends Terminal> terminals = connectable.getTerminals();
+        // check if the connectable are connected with terminal.isConnected()
+        boolean alteastOneIsConnected = false;
```
[the remark this answers is not in the document]
Comment (jamal-khey) [reply src/main/java/org/gridsuite/actions/server/ContingencyListService.java]: done

**Review by AAJELLAL — APPROVED** [review]

**Pull request merged**